server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -0,0 +1,323 @@
+ * created.
+ *
+ */
+public class RemoteFsTranslog extends Translog {
+
+    private final BlobStoreRepository blobStoreRepository;
+    private final TranslogTransferManager translogTransferManager;
+    private final FileTransferTracker fileTransferTracker;
+    private volatile long maxRemoteTranslogGenerationUploaded;
+
+    public RemoteFsTranslog(
+        TranslogConfig config,
+        String translogUUID,
+        TranslogDeletionPolicy deletionPolicy,
+        LongSupplier globalCheckpointSupplier,
+        LongSupplier primaryTermSupplier,
+        LongConsumer persistedSequenceNumberConsumer,
+        BlobStoreRepository blobStoreRepository,
+        ExecutorService executorService
+    ) throws IOException {
+        super(config, translogUUID, deletionPolicy, globalCheckpointSupplier, primaryTermSupplier, persistedSequenceNumberConsumer);
+        this.blobStoreRepository = blobStoreRepository;
+        fileTransferTracker = new FileTransferTracker(shardId);
+        this.translogTransferManager = new TranslogTransferManager(
+            new BlobStoreTransferService(blobStoreRepository.blobStore(), executorService),
+            blobStoreRepository.basePath().add(shardId.getIndex().getUUID()).add(String.valueOf(shardId.id())),
+            fileTransferTracker,
+            fileTransferTracker::exclusionFilter
+        );
+
+        try {
+            final Checkpoint checkpoint = readCheckpoint(location);
+            this.readers.addAll(recoverFromFiles(checkpoint));
+            if (readers.isEmpty()) {
+                throw new IllegalStateException("at least one reader must be recovered");
+            }
+            boolean success = false;
+            current = null;
+            try {
+                current = createWriter(
+                    checkpoint.generation + 1,
+                    getMinFileGeneration(),
+                    checkpoint.globalCheckpoint,
+                    persistedSequenceNumberConsumer
+                );
+                success = true;
+            } finally {
+                // we have to close all the recovered ones otherwise we leak file handles here
+                // for instance if we have a lot of tlog and we can't create the writer we keep
+                // on holding
+                // on to all the uncommitted tlog files if we don't close
+                if (success == false) {
+                    IOUtils.closeWhileHandlingException(readers);
+                }
+            }
+        } catch (Exception e) {
+            // close the opened translog files if we fail to create a new translog...
+            IOUtils.closeWhileHandlingException(current);
+            IOUtils.closeWhileHandlingException(readers);
+            throw e;
+        }
+    }
+
+    /** recover all translog files found on disk */
+    protected ArrayList<TranslogReader> recoverFromFiles(Checkpoint checkpoint) throws IOException {
```

Comment:
If we are recovering from translog files on disk, why do we need a separate implementation here? Can we just put this in `Translog` class?

Reply:
Good catch . Done that now.